### ISSUE TITLE
Feature Request: Easy API to get a talon.skia.Image as a numpy array that can be used by OpenCV

### ISSUE THREAD

**Issue by fmagin** [talonvoice/talon#181]
I am currently using the following code to achieve this myself:

```python
    bitmap = screen.capture(corner.x, corner.y, area_size.x, area_size.y,
                            retina=False).to_bitmap()
    buffer: numpy.ndarray = numpy.frombuffer(bitmap.pixels, dtype=numpy.uint8).copy()
    img = buffer.reshape((-1, area_size.x, 4))
```
this sometimes fails because the buffer isn't long enough, so it would be great if all this logic could be hidden behind some talon API that returns something that can be passed directly to all the OpenCV functions that expect an image.

**Comment by lunixbochs** [comment]
I bet you are taking a screenshot overlapping the edge of the screen, which may return a smaller image than you requested. Try using `bitmap.width` rather than `area_size.x` when reshaping.

**Comment by lunixbochs** [comment]
I just implemented this, seems to round trip fine:

```
im = Image(...)
a = np.array(im)
im = Image.from_array(a)
```

**Comment by lunixbochs** [comment]
shipping in v0.1.4

**Comment by fmagin** [comment]
OpenCV uses BGR while `talon.skia.Image` uses and expects RGB. Currently `Image.from_array` doesn't allow specifying a color space, at least according to the type hints, so this could be a minor possible improvement on that API. Until then, an easy fix is 
`im = skia.Image.from_array(cv2.cvtColor(a, cv2.COLOR_BGR2RGB))`

edit: I messed up the suggested workaround earlier

**Comment by lunixbochs** [comment]
You'll also need to do the reverse if you do np.array(im), and in that case there's nothing I can do on talon's side to switch the color order. Unless there's a numpy convention for BGR, this might be something that should be usually dealt with on the opencv side (so you have the habit of converting in both directions).

Just curious how did you get a BGR image in the first place? Are you using opencv to load an image from disk, or using opencv drawing commands?

**Comment by fmagin** [comment]
I am creating the initial numpy array image with:
```python
img = screen.capture(corner.x, corner.y, area_size.x, area_size.y,
                         retina=False)
bitmap = img.to_bitmap()
buffer: numpy.ndarray = numpy.frombuffer(bitmap.pixels, dtype=numpy.uint8).copy()
img = buffer.reshape((-1, img.width, 4))
img = cv2.cvtColor(img, cv2.COLOR_RGBA2BGR)
```

that `img` than gets processed by a pipeline of OpenCV commands and a copy of the initial image is used to draw the results on for debugging. I am not sure if any of those operation really would break if the image is RGB instead of BGR, I think I just did that because for anything which it is relevant for, OpenCV will assume BGR.

This copy is then used in:
```python
talon_img: skia.Image = skia.Image.from_array(cv2.cvtColor(self.debug_image, cv2.COLOR_BGR2RGB))
src = Rect(0, 0, talon_img.width, talon_img.height)
canvas.draw_image_rect(talon_img, src, canvas.rect)
```
to draw the relevant debug information in a window on the second screen. So this definitely still works well enough, The main annoyance I can think of with this whole colorspace conversion is that matplotlib assumes RGB if you plot a numpy array, while OpenCV assumes BGR if you just pass a numpy array.

You are right that manually converting it with the OpenCV  command isn't a big deal and a useful habit anyway. My point was mostly that `Image.from_pixels` supports a `ColorType` argument, so it seemed somewhat inconsistent for `from_array` to not support this, and worth mentioning in case you agreed.

**Comment by lunixbochs** [comment]
You should be using np.array(im) in latest talon instead of numpy.frombuffer and reshape

**Comment by fmagin** [comment]
Do you mean like this?
```python
    talon_img: skia.Image = screen.capture(corner.x, corner.y, area_size.x, area_size.y,
                         retina=False)
    img = numpy.array(talon_img)
    img = cv2.cvtColor(img, cv2.COLOR_RGBA2BGR)
```
because that doesn't work, and I get a zero dimensional numpy array. I wasn't even aware that this is a thing. Am using this incorrectly or is this possible a mismatch of numpy versions? I installed the numpy version with type stubs earlier, and this might have messed up things in some way, though I am not sure how. I would have expected the Talon upgrade to 0.1.4 to reset that to whatever you used for testing this in the first place. `.talon/bin/pip show numpy` tells me I am on `1.19.5`

edit: Definitely something weird going on with my setup, will open a dedicated issue if I can confirm that this is an issue with a proper install of Talon, or an issue that will potentially crop up with a newer version of numpy.

**Comment by lunixbochs** [comment]
```
>>> im = screen.capture(0, 0, 10, 10)
>>> np.array(im).shape
(10, 10, 4)
>>> np.__version__
'1.20.0rc1'
```

This might be something that will just be fixed once you update to v0.1.5 and get the new Skia version. If you want to debug more, check out `im.__array_interface__`, which for me looks like this:
```
{'shape': (10, 10, 4), 'typestr': '|u1', 'version': 3, 'data': <_cffi_backend.buffer object at 0x10e9bb280>, '__cdata': <cdata 'char[]' owning 400 bytes>}
```

**Comment by fmagin** [comment]
Fascinating bugs I am running into: In my script I take the screenshot with `talon_img: skia.Image = screen.capture(corner.x, corner.y, area_size.x, area_size.y,
                         retina=False)` and accessing `__array_interface__` fails with:
```
2021-01-11 09:31:35 ERROR cb error topic="overlay" cb=<bound method TalonMSER.mser_artist of <user.fmagin.MSER.MSER.TalonMSER object at 0x7fadf81a4f40>>
   19:             talon/canvas.py:16 | 
   18:             talon/canvas.py:152| 
   17: talon/scripting/dispatch.py:133| 
   16: talon/scripting/dispatch.py:126| 
   15:     talon/scripting/rctx.py:201| 
   14: -------------------------------# 'draw' main:draw()
   13:             talon/canvas.py:282| 
   12: talon/scripting/dispatch.py:133| 
   11: talon/scripting/dispatch.py:126| 
   10:     talon/scripting/rctx.py:203| 
    9: -------------------------------# 'overlay' user.fmagin.MSER.MSER:mser_artist()
    8: -------------------------------# stack splice
    7:     talon/scripting/rctx.py:201| 
    6:    user/fmagin/MSER/MSER.py:247| self.grouped_rectangles = list(self.ms..
    5:    user/fmagin/MSER/MSER.py:190| corner, grouped_rectangles, base_img =..
    4:    user/fmagin/MSER/MSER.py:163| corner, screenshot = get_screen_area(p..
    3:    user/fmagin/MSER/MSER.py:64 | print(talon_img.__array_interface__)
    2:         talon/skia/image.py:73 | 
    1:         talon/skia/image.py:171| 
talon.skia.util.SkiaError: pixel read failed
```

But in the REPL where I used `im = screen.capture(0, 0, 10, 10)` it works. Turns out, in my script using `im = screen.capture(0, 0, 10, 10)` or `screen.capture(0, 0, 10, 10, retina=False)` _also_ work and `__array_interface__` correctly returns ` {'shape': (10, 10, 4), 'typestr': '|u1', 'version': 3, 'data': <_cffi_backend.buffer object at 0x7fadf8116bc0>, '__cdata': <cdata 'char[]' owning 400 bytes>}`. Do you have some idea why this might be? It might be that I am passing `float` as x,y,width,height but this is explicitly what the type annotations support.
